File: apps/passenger-android/src/main/kotlin/lk/mageride/passenger/live/LiveHubInbox.kt
package lk.mageride.passenger.live

import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import lk.mageride.shared.data.api.query.QueryApi
import lk.mageride.shared.data.models.GeoPoint
import lk.mageride.shared.domain.geo.H3Cell
import lk.mageride.shared.domain.geo.H3Grid
import lk.mageride.shared.realtime.LiveHub
import lk.mageride.shared.realtime.ShareRevoked
import lk.mageride.shared.realtime.VehicleFrame
import lk.mageride.shared.realtime.VehicleRemoved
import lk.mageride.shared.serialization.MageRideJson

/**
 * Everything the server says, turned into what the map shows.
 *
 * The socket carries deltas and `GET /v1/nearby` carries snapshots, and both end up in the same
 * [LiveVehicleStore] — which is why the resync lives here rather than beside the connection. From
 * a screen's point of view there is one answer to "what is nearby", and it does not matter which
 * transport most recently updated it.
 *
 * **Payloads arrive as JSON text and are decoded by `MageRideJson`.** The SignalR Java client's
 * hub protocol is Gson, and Gson binds an enum by its Kotlin `name()`; C012's enums carry
 * `@SerialName` wire spellings that differ (`three_wheeler`, not `THREE_WHEELER`). Decoding with
 * the platform's own `Json` is what lets the socket and the REST surface share one set of models,
 * which is what `signalr-hub.md` §3 asks for. See [SignalRLiveHubTransport].
 *
 * @param query query-svc, for the snapshot half of D6' §5.4's recovery.
 * @param grid The platform H3 engine, for pruning vehicles in cells the client has left.
 */
internal class LiveHubInbox(private val query: QueryApi, private val grid: H3Grid) {

    private val store = LiveVehicleStore()
    private val mutex = Mutex()

    private val mutableVehicles = MutableStateFlow<List<VehicleFrame>>(emptyList())
    private val mutableEvents = MutableSharedFlow<LiveEvent>(extraBufferCapacity = EVENT_BUFFER)

    /** What is on the map. Replaced wholesale per batch — see [LiveVehicleStore]. */
    val vehicles: StateFlow<List<VehicleFrame>> = mutableVehicles.asStateFlow()

    /** The four directed payloads. See [LiveEvent]. */
    val events: SharedFlow<LiveEvent> = mutableEvents.asSharedFlow()

    /**
     * One hub event.
     *
     * Every decode is wrapped, because a malformed or unrecognised payload must not take down the
     * callback that every other event also arrives on. `MageRideJson` has `ignoreUnknownKeys`, so
     * a field added server-side is not malformed — what this actually guards is a contract change
     * big enough to change a *shape*, which is a deploy problem rather than a reason to lose the
     * map until the app is restarted.
     */
    suspend fun onEvent(name: String, payload: String) {
        runCatching { dispatch(name, payload) }
    }

    /**
     * The snapshot half of the recovery — *"the socket carries deltas, the REST read carries the
     * state"*.
     *
     * A failure is swallowed: the socket is already up and the next per-cell batch is at most
     * eight seconds away, so a failed resync costs freshness rather than the map. The radius is
     * the R-06 view's own ~3 km, which is what the nineteen cells cover.
     */
    suspend fun resync(around: GeoPoint) {
        runCatching { query.getNearbyVehicles(lat = around.lat, lng = around.lng, radiusMetres = SNAPSHOT_RADIUS_M) }
            .onSuccess { response ->
                mutex.withLock { store.onSnapshot(response.vehicles) }
                publish()
            }
    }

    /** Drops every vehicle whose own res-7 cell is no longer subscribed. See [LiveVehicleStore]. */
    suspend fun retainCells(cells: Set<H3Cell>) {
        if (mutex.withLock { store.retainCells(cells, grid) }) publish()
    }

    /** Forgets the map. The passenger signed out, or the app is going. */
    suspend fun clear() {
        mutex.withLock { store.clear() }
        publish()
    }

    private suspend fun dispatch(name: String, payload: String) {
        when (name) {
            LiveHub.Event.VEHICLE_POSITIONS -> {
                val frames = MageRideJson.decodeFromString<List<VehicleFrame>>(payload)
                if (mutex.withLock { store.onPositions(frames) }) publish()
            }

            LiveHub.Event.VEHICLE_REMOVED -> {
                val removed = MageRideJson.decodeFromString<VehicleRemoved>(payload)
                if (mutex.withLock { store.onRemoved(removed) }) publish()
            }

            LiveHub.Event.SHARE_REVOKED -> {
                val revoked = MageRideJson.decodeFromString<ShareRevoked>(payload)
                if (mutex.withLock { store.onShareRevoked(revoked) }) publish()
            }

            LiveHub.Event.RIDE_STATE_CHANGED ->
                mutableEvents.emit(LiveEvent.RideState(MageRideJson.decodeFromString(payload)))

            LiveHub.Event.DRIVER_POSITION ->
                mutableEvents.emit(LiveEvent.DriverMoved(MageRideJson.decodeFromString(payload)))

            LiveHub.Event.LOCATION_REQUEST_RESOLVED ->
                mutableEvents.emit(LiveEvent.LocationRequest(MageRideJson.decodeFromString(payload)))

            LiveHub.Event.PACKAGE_STATUS ->
                mutableEvents.emit(LiveEvent.PackageStatus(MageRideJson.decodeFromString(payload)))
        }
    }

    private suspend fun publish() {
        mutableVehicles.value = mutex.withLock { store.snapshot() }
    }

    internal companion object {

        /**
         * The `GET /v1/nearby` radius, in metres.
         *
         * The R-06 view is res-7 + `ring(2)`, which reaches 2.8–3.3 km; 3 km is inside the
         * contract's 100–20 000 m bounds and matches what the nineteen cells actually cover. The
         * cells remain the subscription — this is only what the resync asks for.
         */
        const val SNAPSHOT_RADIUS_M: Int = 3_000

        /**
         * How many directed events may queue for a screen that is slow to collect.
         *
         * `emit` suspends rather than overflowing, so nothing is dropped; the buffer is what keeps
         * a `RideStateChanged` burst mid-accept from serialising against a view model's fold.
         */
        private const val EVENT_BUFFER = 8
    }
}

File: apps/passenger-ios/PassengerApp/Live/LiveHubInbox.swift
import Foundation
import MageRideShared

/// The four directed payloads a screen subscribes to.
///
/// `VehiclePositions`, `VehicleRemoved` and `ShareRevoked` are deliberately absent: all three are
/// about *what is on the map*, and the map reads ``PassengerLiveMap/vehicles`` rather than replaying
/// a stream to rebuild a set the store already holds. `ShareRevoked` also reaches a screen through
/// the vehicle disappearing, which is the only consequence it has.
enum LiveEvent {

    /// A ride-aggregate transition — SCR-PI-014/015/017/020's own state.
    case rideState(RideStateChanged)

    /// The assigned driver's live position (US-6A.12) — SCR-PI-015's moving marker.
    case driverMoved(DriverPosition)

    /// The proxy round-trip resolving (P-02, P-13) — SCR-PI-010b's pin auto-filling.
    case locationRequest(LocationRequestResolved)

    /// Package handoff progress (US-20.7) — SCR-PI-020/021's four-step bar.
    case packageStatus(PackageStatusChanged)
}

/// The snapshot half of D6' §5.4's recovery, as one method.
///
/// **A seam rather than `QueryApi` itself, and the reason is the bridge.** `QueryApi` is a Kotlin
/// interface with nine `suspend` methods; implementing one in Swift means implementing all nine, and
/// implementing an *exported suspend* function from Swift means matching whatever completion-handler
/// selector Kotlin/Native generated for it — a shape that is the compiler's business rather than
/// this codebase's. One method the app declares is testable with four lines and cannot drift.
///
/// It is also the narrower truth: the live plane needs exactly one read out of query-svc, and a
/// protocol that says so is a protocol nobody can quietly add a second call to.
protocol NearbySnapshots: AnyObject {

    /// `GET /v1/nearby` — what is visible around a point, right now.
    ///
    /// Answers `nil` on any failure. The socket is already up and the next per-cell batch is at most
    /// eight seconds away, so a failed resync costs freshness rather than the map.
    func nearby(around point: GeoPoint, radiusMetres: Int) async -> [NearbyVehicle]?
}

/// ``NearbySnapshots`` over the real client.
///
/// The whole of what it adds is the two things a Swift call site cannot leave out: **every parameter
/// is passed**, because a Kotlin default argument does not survive the Objective-C export, and the
/// throw is swallowed here rather than at the caller.
///
/// `types` and `modes` are `nil` on purpose. The *platform's* visibility rules are the server's
/// (entitlement, freshness, engagement) and SCR-PI-006's filter is the passenger's — folding the
/// second into this call would make a Mode B vehicle they have no grant for indistinguishable from
/// one they simply switched off. C078 makes the same argument on the Android side.
final class ApiNearbySnapshots: NearbySnapshots {

    private let query: QueryApi

    init(query: QueryApi) {
        self.query = query
    }

    func nearby(around point: GeoPoint, radiusMetres: Int) async -> [NearbyVehicle]? {
        try? await query.getNearbyVehicles(
            lat: point.lat,
            lng: point.lng,
            radiusMetres: KotlinInt(int: Int32(radiusMetres)),
            types: nil,
            modes: nil
        ).vehicles
    }
}

/// Everything the server says, turned into what the map shows.
///
/// The socket carries deltas and `GET /v1/nearby` carries snapshots, and both end up in the same
/// ``LiveVehicleStore`` — which is why the resync lives here rather than beside the connection. From
/// a screen's point of view there is one answer to "what is nearby", and it does not matter which
/// transport most recently updated it.
///
/// **Payloads arrive as JSON text and are decoded by `:shared`.** `IosLiveHubPayloadsKt` is the
/// door; it exists because `MageRideJson.decodeFromString<T>()` is `inline` + `reified` and is not
/// exported to Objective-C at all, and because a `Codable` mirror of `VehicleFrame` in this target
/// is what `signalr-hub.md` §3 forbids. See ``SignalRLiveHubTransport``.
///
/// An `actor` because the store it drives is a value with no lock of its own and the events arrive
/// on the client's own callback queue.
actor LiveHubInbox {

    private let snapshots: NearbySnapshots
    private let grid: H3Grid
    private let contract = IosLiveHub()
    private var store = LiveVehicleStore()

    /// Called with the whole visible set whenever it changes — one update per batch, not one per
    /// vehicle. ``PassengerLiveMap`` hops it to the main actor and publishes it.
    private let onVehicles: ([VehicleFrame]) -> Void

    /// Called with each directed payload, in arrival order.
    private let onEvent: (LiveEvent) -> Void

    init(
        snapshots: NearbySnapshots,
        grid: H3Grid,
        onVehicles: @escaping ([VehicleFrame]) -> Void,
        onEvent: @escaping (LiveEvent) -> Void
    ) {
        self.snapshots = snapshots
        self.grid = grid
        self.onVehicles = onVehicles
        self.onEvent = onEvent
    }

    /// One hub event.
    ///
    /// A payload that does not decode is dropped rather than thrown: it must not take down the
    /// callback every other event also arrives on, and on this platform the decoders answer `nil`
    /// rather than throwing for a second reason — an exception out of a non-suspend Kotlin function
    /// crosses as an uncaught Objective-C exception, which Swift cannot catch (the C091 finding).
    func receive(event name: String, payload: String) {
        switch name {
        case contract.eventVehiclePositions:
            guard let frames = IosLiveHubPayloadsKt.decodeVehicleFrames(json: payload) else { return }
            publishIf(store.onPositions(frames))

        case contract.eventVehicleRemoved:
            guard let removed = IosLiveHubPayloadsKt.decodeVehicleRemoved(json: payload) else { return }
            publishIf(store.onRemoved(removed))

        case contract.eventShareRevoked:
            guard let revoked = IosLiveHubPayloadsKt.decodeShareRevoked(json: payload) else { return }
            publishIf(store.onShareRevoked(revoked))

        case contract.eventRideStateChanged:
            guard let value = IosLiveHubPayloadsKt.decodeRideStateChanged(json: payload) else { return }
            onEvent(.rideState(value))

        case contract.eventDriverPosition:
            guard let value = IosLiveHubPayloadsKt.decodeDriverPosition(json: payload) else { return }
            onEvent(.driverMoved(value))

        case contract.eventLocationRequestResolved:
            guard let value = IosLiveHubPayloadsKt.decodeLocationRequestResolved(json: payload) else { return }
            onEvent(.locationRequest(value))

        case contract.eventPackageStatus:
            guard let value = IosLiveHubPayloadsKt.decodePackageStatusChanged(json: payload) else { return }
            onEvent(.packageStatus(value))

        default:
            // An event this build does not know. `signalr-hub.md` §3 has seven; a later one arrives
            // here and is ignored, which is what a client should do with news it has no screen for.
            return
        }
    }

    /// The snapshot half of the recovery — *"the socket carries deltas, the REST read carries the
    /// state"*.
    ///
    /// A failure is swallowed: the socket is already up and the next per-cell batch is at most eight
    /// seconds away, so a failed resync costs freshness rather than the map. The radius is the R-06
    /// view's own ~3 km, which is what the nineteen cells cover.
    func resync(around point: GeoPoint) async {
        guard let vehicles = await snapshots.nearby(
            around: point,
            radiusMetres: LiveHubInbox.snapshotRadiusMetres
        ) else { return }

        publishIf(store.onSnapshot(vehicles))
    }

    /// Drops every vehicle whose own res-7 cell is no longer subscribed. See ``LiveVehicleStore``.
    func retain(cells: Set<H3Cell>) {
        publishIf(store.retain(cells: cells, grid: grid))
    }

    /// Drops one vehicle the passenger just unsubscribed from (AL-25).
    func drop(vehicleId: String) {
        publishIf(store.drop(vehicleId))
    }

    /// Forgets the map. The passenger signed out, or the app is going.
    func clear() {
        _ = store.clear()
        onVehicles(store.snapshot())
    }

    private func publishIf(_ changed: Bool) {
        guard changed else { return }
        onVehicles(store.snapshot())
    }

    /// The `GET /v1/nearby` radius, in metres.
    ///
    /// The R-06 view is res-7 + `ring(2)`, which reaches 2.8–3.3 km; 3 km is inside the contract's
    /// 100–20 000 m bounds and matches what the nineteen cells actually cover. The cells remain the
    /// subscription — this is only what the resync asks for.
    static let snapshotRadiusMetres = 3_000
}
